lms/djangoapps/bulk_email/messages.py

```diff
@@ -0,0 +1,84 @@
+"""
+Module to define email message related classes and methods
+"""
+from abc import ABC, abstractmethod
+
+from django.contrib.auth import get_user_model
+from django.core.mail import EmailMultiAlternatives
+from edx_ace import ace
+from edx_ace.recipient import Recipient
+
+from lms.djangoapps.bulk_email.message_types import BulkEmail
+from openedx.core.lib.celery.task_utils import emulate_http_request
+
+User = get_user_model()
+
+
+class CourseEmailMessage(ABC):
+    """
+    Abstract base class for course email messages
+    """
+
+    @abstractmethod
+    def send(self):
+        """
+        Triggers sending of email message
+        """
+
+
+class DjangoEmail(CourseEmailMessage):
+    """
+    Email message class to send email directly using django mail API.
+    """
+    def __init__(self, connection, course_email, email_context):
+        """
+        Construct message content using course_email model and context
+        """
+        self.connection = connection
+        template_context = email_context.copy()
+        # use the CourseEmailTemplate that was associated with the CourseEmail
+        course_email_template = course_email.get_template()
+
+        plaintext_msg = course_email_template.render_plaintext(course_email.text_message, template_context)
+        html_msg = course_email_template.render_htmltext(course_email.html_message, template_context)
+
+        # Create email:
+        message = EmailMultiAlternatives(
+            course_email.subject,
+            plaintext_msg,
+            email_context['from_address'],
+            [email_context['email']]
+        )
+        message.attach_alternative(html_msg, 'text/html')
+        self.message = message
+
+    def send(self):
+        """
+        send email using already opened connection
+        """
+        self.connection.send_messages([self.message])
+
+
+class ACEEmail(CourseEmailMessage):
```

Comment:
I'm curious about the use of the abstract base class `CourseEmailMessage`. Is there any reason we went this route?

I'm no expert on the use of `edx-ace` but the convention I am _used_ to seeing is the message class inheriting from the `MessageType` class (https://github.com/openedx/edx-ace/blob/958255a3b4b4d1f58de92f8544867372af99254b/edx_ace/message.py#L149).

Most messages in the LMS inherit from `BaseMessageType` (https://github.com/openedx/edx-platform/blob/a94d5e1ae3efa3d629cd547e5b19fee1cc22a9b0/openedx/core/djangoapps/ace_common/message.py#L11 which inherits from MessageType).

I'm also used to seeing this pattern in the Credentials IDA as well (https://github.com/openedx/credentials/blob/f9c34a5b95d03d08f25f3f4d9ca6512c98d7b2c4/credentials/apps/records/messages.py#L7).